### PR TITLE
[local_auth] Allow device authentication (pin/pattern/passcode)

### PR DESCRIPTION
**Pull request by kennethj**
## Description

Added `LocalAuthentication().authenticate()` method that allows users to authenticate using biometric (fingerprint, face) or device authentication (pin, pattern, passcode).

## Screenshots
ios-auth
Screen Shot 2020-01-31 at 8 29 51 AM
## Related Issues

This resolves the feature request on https://github.com/flutter/flutter/issues/49703

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by jitenders859** [comment]
on Android Can see the use password button but when clicking on it it closes the dialog and shows this message in console

FingerprintDialogFrag( 4198): Failed to check device credential. Parent handler not found.

**Comment by kennethj** [comment]
@jitenders859 Is that on a physical device or with on an emulator?
I was having issues running this on android emulators, but it appears to work on physical devices. 
After a little research, I found there may be `KeyguardManager#isDeviceSecure()` check that could be added to avoid the error.

**Comment by jitenders859** [comment]
@kennethj I am using real android device and 
also can you show how to implement KeyguardManager#isDeviceSecure()

also in file

https://android.googlesource.com/platform/frameworks/support/+/refs/heads/androidx-recyclerview-release/biometric/src/main/java/androidx/biometric/Utils.java
From the source code we can see at line 82 that if the activity type is not of DeviceCredentialHandlerActivity it will fail

**Comment by jitenders859** [comment]
@kennethj 

ezgif-3-b71f99e54d14

In the above gif when pressed on authenticated the popup opens two times and when pressed on use password it closes on reopening can't see fingerprint icon and on authenticating with fingerprint didn't authenticate. after that opening second time it authenticates using fingerprint.

also is there any problem with sdk version below 28

**Comment by kennethj** [comment]
@jitenders859 
This does appear to be an issue with API 28 and below. This feature only seems to work as intended on API 29 (Android 10). I am going to try to find time today to look into it more.

**Comment by apgapg** [comment]
@kennethj In Android 6.0 with non-fingerprint device it doesnt allow to change verification method.
Device: **Samsung J2 Pro, SM-J210F android 6.0.1  without fingerprint**

Here's my implementation:
```
  local_auth:
    git:
      url: https://github.com/kennethj/flutter-plugins
      ref: feature/local_auth
      path: packages/local_auth

```
my expectation is to have pattern verification also which works fine in higher android version. Only in 6.0 it fails.

IMG-20200603-WA0005

Screenshot_2020-06-03-15-45-54-822_com workozy android demo

**Comment by apgapg** [comment]
Any progress in this PR?
We really need pattern lock verification because some devices dont have fingerprint.

**Comment by SergejSachs** [comment]
yea thats true, device password auth would be so good as a fallback

**Comment by kdhanji87** [comment]
Any progress in this PR?
We really need pattern lock verification because some devices dont have fingerprint would be a useful fallback

**Comment by duncwillz** [comment]
@kennethj LGTM. @softmarshmallow @mehmetf Could you please help get this PR merged?. It wouldn't be worth reinventing the wheel locally. Thank you.

**Comment by Zelosys** [comment]
Hello
Any ETA on this PR?

It's a necessary feature for our App, and we don't want to reinvent the Wheel and do it manual, when the plugin can do this soon by its own :)

**Review by softmarshmallow — APPROVED** [review]

**Comment by SergejSachs** [comment]
Any update on this?

**Comment by Zelosys** [comment]
It looks like after three months it's time to annoy the Google staff again.😀
Any updates on this? @mehmetf 
Is there anything we can do to speed this up and get it in?

**Comment by mehmetf** [comment]
Is this PR actively maintained? It needs to be rebased. Please ping me and I will do a pass after someone claims ownership.

@Zelosys @kennethj

**Comment by kennethj** [comment]
@mehmetf I will probably have some time in the next two weeks or so.

**Comment by kennethj** [comment]
@mehmetf I think this is ready for review.

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java
Comment (mehmetf) [inline packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java:92]: You can pass  `allowCredentials` to  `setDeviceCredentialAllowed` call and get rid  of the conditional.

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java
Comment (mehmetf) [inline packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java:142]: These  strings should come from Dart just like the rest of the status strings used below.

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java
Comment (mehmetf) [inline packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java:234]: Do we know why? Do the handlers below fire twice for some reason?

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
Comment (mehmetf) [inline packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:318]: Can you please move this definition up, before it is used for the first time?

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
Comment (mehmetf) [inline packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:90]: There's no need for this constructor.  You can do all this in the registerWith() call.

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
Comment (mehmetf) [inline packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:54]: Should  you ever be here? I don't see you checking  for null in everywhere getActivity is called  below.

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
Comment (mehmetf) [inline packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:292]: If these services can be grabbed via ApplicationContext, prefer doing that and save  them as instance members to the plugin class. That way you don't have to get them with every call and eliminate the possibility that getActivity could be null.

File: packages/local_auth/example/lib/main.dart
Comment (mehmetf) [inline packages/local_auth/example/lib/main.dart:25]: Prefer  initializing variables rather than leaving them nullable.

File: packages/local_auth/lib/auth_strings.dart
Comment (mehmetf) [inline packages/local_auth/lib/auth_strings.dart:18]: This is a breaking change. Could you please add it to the CHANGELOG?

File: packages/local_auth/lib/local_auth.dart
Comment (mehmetf) [inline packages/local_auth/lib/local_auth.dart:69]: This should not return a nullable type.

File: packages/local_auth/lib/local_auth.dart
Comment (mehmetf) [inline packages/local_auth/lib/local_auth.dart:138]: Since you are making a breaking change already due to message strings, why not rename the method `authenticateWithBiometrics` as `authenticate` instead of having a near duplicate method?

That would leave the code much simpler.

**Review by mehmetf — COMMENTED** [review]
Thanks for the contribution (and especially tests). There's additional work required before we can merge this into the repo.

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java
Comment (kennethj) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java:92]: There was an issue with setting both `setNegativeButtonText` and `setDeviceCredentialAllowed`. 


From android docs:
> Note that enabling this option replaces the negative button on the prompt with one that allows the user to authenticate with their device credential, making it an error to call setNegativeButton(java.lang.CharSequence, java.util.concurrent.Executor, android.content.DialogInterface.OnClickListener).

I will refactor to try to make it more obvious what is happening.

File: packages/local_auth/example/lib/main.dart
Comment (kennethj) [reply packages/local_auth/example/lib/main.dart:25]: I'll add an enum to avoid the `null`

File: packages/local_auth/lib/local_auth.dart
Comment (kennethj) [reply packages/local_auth/lib/local_auth.dart:138]: I wanted to keep the original behavior intact of using biometrics only with `authenticateWithBiometrics` and have `authenticate` use any means available, then perhaps a `authenticateWithCredentials` at some point.

If we remove `authenticateWithBiometrics`, I would want to add a way to keep the original biometric only behavior. Maybe a `bool biometricOnly = false,` parameter on `authenticate`?

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java
Comment (kennethj) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java]: updated

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java
Comment (kennethj) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java]: Added to `AndroidAuthMessages` class

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
Comment (kennethj) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java]: moved up

File: packages/local_auth/example/lib/main.dart
Comment (kennethj) [reply packages/local_auth/example/lib/main.dart]: Updated to use the `enum` `_SupportState` instead of a nullable `bool`

File: packages/local_auth/lib/local_auth.dart
Comment (kennethj) [reply packages/local_auth/lib/local_auth.dart]: Updated to return `Future<bool>` instead of `Future<bool?>`

**Comment by kennethj** [comment]
@mehmetf, thanks for reviewing that so quickly!

I will test these changes on a few emulators and finish updates based on your review tomorrow.

File: packages/local_auth/lib/local_auth.dart
Comment (mehmetf) [reply packages/local_auth/lib/local_auth.dart]: Yes, a `biometricOnly` parameter sounds good.

I like your original idea  as well (having separate methods) however the implementation (and API) is largely shared between these methods and you would need to make sure a feature added to `authenticate` is also added to all of its 
 variants (`authenticateWithBiometrics` etc).

You can reduce the implementation of `authenticateWithBiometrics` to just calling `authenticate` with `biometricOnly` set to true. However, I would rather keep the API clean and simple.

File: packages/local_auth/lib/auth_strings.dart
Comment (kennethj) [reply packages/local_auth/lib/auth_strings.dart:18]: Noted to changelog.

File: packages/local_auth/lib/local_auth.dart
Comment (kennethj) [reply packages/local_auth/lib/local_auth.dart]: Added a `biometricOnly` parameter to `authenticate`. `authenticateWithBiometrics` now calls `authenticate` with `biometricOnly: true`

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java
Comment (kennethj) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java:228]: I am not sure. This happened a while ago and don't remember the specifics and I have not been able to reproduce the issue in emulators today.

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
Comment (kennethj) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java]: Moved code to `registerWith()` method and deleted extra constructor.

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java
Comment (mehmetf) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java:228]: Let's remove it if it cannot be reproduced.

**Comment by mehmetf** [comment]
Thanks for making all these changes. Let me know when it is ready for  the second round!

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java
Comment (kennethj) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java]: removed

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
Comment (kennethj) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java]: Removed `getActivity()`. Setting services and activity as instances variables when plugin registered and on attached to activity.

**Comment by kennethj** [comment]
@mehmetf 
I believe I addressed all the comments in the first review and ready for a second look.

Also, I am not sure what's up with the failing CI build `build-apks+java-test+firebase-test-lab CHANNEL:stable PLUGIN_SHARDING:--shardIndex 0 --shardCount 4`. I started updating gradle versions, but that started to become a rabbit hole.

**Comment by zigapovhe** [comment]
> @mehmetf
> I believe I addressed all the comments in the first review and ready for a second look.
> 
> Also, I am not sure what's up with the failing CI build `build-apks+java-test+firebase-test-lab CHANNEL:stable PLUGIN_SHARDING:--shardIndex 0 --shardCount 4`. I started updating gradle versions, but that started to become a rabbit hole.

@kennethj Did you enable  'android.useAndroidX' property to true? Set this property to true in the gradle.properties file and retry.

File: packages/local_auth/CHANGELOG.md
Comment (mehmetf) [inline packages/local_auth/CHANGELOG.md:40]: We are not going to publish this as all changes are now done in the null-safety version. Please move to the CHANGELOG description above.

File: packages/local_auth/CHANGELOG.md
Comment (mehmetf) [inline packages/local_auth/CHANGELOG.md:75]: Remove these extra newlines. It might be due to formatting done in your editor?

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
Comment (mehmetf) [inline packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:98]: This is dangerous. I assume you are saving result because some operation below can respond asynchronously via `onPermissionResult`, `onActivityResult` or some such. You should save the result object on a per-method basis right before you invoke the `startActivityForResult`. Otherwise, another method call can override this result object and you end up calling `result` methods more than once.

This also gives you the chance to reject a second call to the async method. e.g. if you are still awaiting the result of some `foo` method and user calls `foo` again, you can ignore it and instead wait for the `Result  fooResult` to be fulfilled.

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
Comment (mehmetf) [inline packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:101]: `this.` qualifier is unnecessary (ditto below).

File: packages/local_auth/lib/local_auth.dart
Comment (mehmetf) [inline packages/local_auth/lib/local_auth.dart:76]: I would strongly recommend deprecating this method and then removing it sometime in the future.

Please put a `@deprecated` qualifier to it, remove its dartdoc replacing it with "This method is deprecated, please use `authenticate` with `biometricOnly: true`."

File: packages/integration_test/android/build.gradle
Comment (mehmetf) [inline packages/integration_test/android/build.gradle:11]: I am not entirely sure how version resolving for gradle works for this testbed. I will get someone from Flutter core team to review these particular changes. That's likely what's causing your build failure.

Comment (mehmetf) [reply packages/integration_test/android/build.gradle:11]: @cyanglaz Could you please take a look at the build file changes in this PR? What is the proper way to up the gradle version?

File: packages/local_auth/CHANGELOG.md
Comment (kennethj) [reply packages/local_auth/CHANGELOG.md]: Moved.

File: packages/local_auth/CHANGELOG.md
Comment (kennethj) [reply packages/local_auth/CHANGELOG.md]: extra newlines removed

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
Comment (kennethj) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java]: Updated

File: packages/local_auth/lib/local_auth.dart
Comment (kennethj) [reply packages/local_auth/lib/local_auth.dart:43]: Added `@Deprecated` annotation and comment docs

File: packages/integration_test/android/build.gradle
Comment (kennethj) [reply packages/integration_test/android/build.gradle]: I revert the changes to the `integration_test` package in this PR. This appears to prevent those failed builds.

File: packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java
Comment (mehmetf) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java]: I think this is now the only blocker before I LGTM this PR.

Comment (kennethj) [reply packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java]: Removed class level `result` field. Added`lockRequestResult` field that is only set when lock request activity started. This field `lockRequestResult` is only used in activity result delegate function.

**Review by mehmetf — APPROVED** [review]

**Comment by mehmetf** [comment]
Published as 1.1.0-nullsafety.

File: packages/local_auth/ios/Classes/FLTLocalAuthPlugin.m
Comment (stuartmorgan) [inline packages/local_auth/ios/Classes/FLTLocalAuthPlugin.m:144]: Why is fallback being disabled here? That seems undesirable for the non-biometric-only case.

Comment (mtskf) [reply packages/local_auth/ios/Classes/FLTLocalAuthPlugin.m:144]: Please escalate this!

**Pull request merged**
